[Style] Convert grid-template-areas to use strong style types

weinig opened the pull request:
#### 2d2562d50a84003b593458a656b99c4c3969942f
<pre>
[Style] Convert grid-template-areas to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=295714">https://bugs.webkit.org/show_bug.cgi?id=295714</a>

Reviewed by Darin Adler.

Bundles up the values for the `grid-template-areas` property into a new
`Style::GridTemplateAreas` type.

Also modernizes the CSS side a, bundling up the map and row/column count
values into a struct.

* Source/WebCore/Headers.cmake:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
* Source/WebCore/css/CSSGridTemplateAreasValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
* Source/WebCore/css/parser/CSSPropertyParserCustom.h:
* Source/WebCore/css/values/grid/CSSGridNamedAreaMap.cpp: Added.
* Source/WebCore/css/values/grid/CSSGridNamedAreaMap.h: Added.
* Source/WebCore/css/values/grid/CSSGridTemplateAreas.cpp: Added.
* Source/WebCore/css/values/grid/CSSGridTemplateAreas.h: Added.
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/style/GridArea.h:
* Source/WebCore/rendering/style/GridPositionsResolver.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleOrderedNamedLinesCollector.h:
* Source/WebCore/style/values/grid/StyleGridNamedLinesMap.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridNamedLinesMap.h: Added.
* Source/WebCore/style/values/grid/StyleGridOrderedNamedLinesMap.h: Added.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.h: Added.

Canonical link: <a href="https://commits.webkit.org/297304@main">https://commits.webkit.org/297304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/549b53f86b2ea490c7144db2f8821eaa3f502115

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111212 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21392 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117243 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61480 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113174 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39460 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84533 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100172 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64979 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24547 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18313 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61063 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94583 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120305 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28430 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93463 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38637 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93288 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23781 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38380 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16188 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34246 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38150 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43627 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37815 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41148 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39517 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
```diff
@@ -32,49 +32,43 @@
 
-Ref<CSSGridTemplateAreasValue> CSSGridTemplateAreasValue::create(NamedGridAreaMap map, size_t rowCount, size_t columnCount)
+CSSGridTemplateAreasValue::CSSGridTemplateAreasValue(const CSS::GridTemplateAreas& areas)
+    : CSSValue(ClassType::GridTemplateAreas)
```
Would this be efficient enough?
```suggestion
CSSGridTemplateAreasValue::CSSGridTemplateAreasValue(const CSS::GridTemplateAreas& areas)
    : CSSGridTemplateAreasValue { CSS::GridTemplateAreas { areas } }
{
}
```
I like the idea of not needing to write the assertions twice.

darinadler commented on Source/WebCore/css/CSSGridTemplateAreasValue.h:
```diff
@@ -31,31 +31,28 @@
 private:
-    CSSGridTemplateAreasValue(NamedGridAreaMap, size_t rowCount, size_t columnCount);
+    CSSGridTemplateAreasValue(CSS::GridTemplateAreas&&);
+    CSSGridTemplateAreasValue(const CSS::GridTemplateAreas&);
```
```suggestion
    explicit CSSGridTemplateAreasValue(CSS::GridTemplateAreas&&);
    explicit CSSGridTemplateAreasValue(const CSS::GridTemplateAreas&);
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
```diff
@@ -148,10 +152,10 @@ bool parseGridTemplateAreasRow(StringView gridRowNames, NamedGridAreaMap& gridAr
         currentColumn = lookAheadColumn - 1;
     }
 
+    gridAreaMap.rowCount++;
```
Our usual style is this:
```suggestion
    ++gridAreaMap.rowCount;
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
```diff
@@ -49,7 +48,7 @@ enum class AllowEmpty : bool { No, Yes };
 
 bool isGridBreadthIdent(CSSValueID);
-bool parseGridTemplateAreasRow(StringView gridRowNames, NamedGridAreaMap&, size_t rowCount, size_t& columnCount);
+bool parseGridTemplateAreasRow(StringView gridRowNames, CSS::GridNamedAreaMap&);
```
In future could it be this?
```suggestion
std::optional<CSS::GridNamedAreaMap> parseGridTemplateAreasRow(StringView);
```

weinig replied:
Not exactly, since it adds to existing maps

darinadler replied:
The name didn’t make that clear to me. Usually a function named “parse” converts an unparsed form to a parsed form or fails. Should we create a naming scheme for “parse and update data structure” as opposed to “parse and create object for the parsed result"? I typically use the word “update” somewhere in the name of something like that. Or maybe argument types make that obvious to people other than me? Or maybe we’d like a different name, but can’t think of a compelling one?

weinig replied:
Yeah, that whole thing is not clear at all.

A caller currently looks like this:

```c++
    CSS::GridNamedAreaMap map;
    while (range.peek().type() == StringToken) {
        if (!parseGridTemplateAreasRow(range.consumeIncludingWhitespace().value(), map))
            return nullptr;
    }
```

I think the right way to handle this is probably to break it up.

The actual parsing happens in a function called `parseGridTemplateAreasColumnNames()`. So, do something like:

```c++
    CSS::GridNamedAreaMap map;
    while (range.peek().type() == StringToken) {
        auto columnNames = parseGridTemplateAreasColumnNames(names);
        if (!columnNames)
            return nullptr;

        if (!updateNamedAreaMapWithRow(map, *columnNames))
            return nullptr;
    }
```

or if we want to add a member function:

```c++
    CSS::GridNamedAreaMap map;
    while (range.peek().type() == StringToken) {
        auto columnNames = parseGridTemplateAreasColumnNames(names);
        if (!columnNames)
            return nullptr;

        if (!map.update(*columnNames))
            return nullptr;
    }
```

darinadler replied:
Well if that function and data structure already exists, then this seems like an straightforward and helpful refactor! I didn’t realize there already was a vector of parsed names that was inside the function.

No reason to do this in this pull request, though, I suppose.

darinadler commented on Source/WebCore/css/values/grid/CSSGridNamedAreaMap.cpp:
```diff
@@ -0,0 +1,72 @@
+
+static void serializeGridNamedAreaMapPosition(StringBuilder& builder, const GridNamedAreaMap::Map& map, size_t row, size_t column)
+{
+    UncheckedKeyHashSet<String> candidates;
```
We are aiming to use UncheckedKeyHashSet only where it’s required for a measurable performance gain.
```suggestion
    HashSet<String> candidates;
```

darinadler commented on Source/WebCore/css/values/grid/CSSGridNamedAreaMap.cpp:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+TextStream& operator<<(TextStream& ts, const GridNamedAreaMap& value)
+{
+    return ts << serializationForCSS(defaultSerializationContext(), value);
+}
```
Is `ts` better than `stream`?
```suggestion
TextStream& operator<<(TextStream& stream, const GridNamedAreaMap& value)
{
    return stream << serializationForCSS(defaultSerializationContext(), value);
}
```

weinig replied:
That's begging for big `do-webcore-rename`.

darinadler commented on Source/WebCore/css/values/grid/CSSGridNamedAreaMap.h:
```diff
@@ -0,0 +1,51 @@
+
+// Parsed representation of the `<string>+` of <'grid-template-areas'>.
+struct GridNamedAreaMap {
+    using Map = UncheckedKeyHashMap<String, GridArea>;
```
Do we need to use `UncheckedKeyHashMap`?
```suggestion
    using Map = HashMap<String, GridArea>;
```
It is OK that we lose the order of the areas, since we keep them in a map?

weinig replied:
Yeah, its a map of "areas", and you can reconstruct the original strings by looking through all the areas.

darinadler commented on Source/WebCore/css/values/grid/CSSGridNamedAreaMap.h:
```diff
@@ -0,0 +1,51 @@
+    using Map = UncheckedKeyHashMap<String, GridArea>;
+
+    Map map;
+    size_t rowCount { 0 };
+    size_t columnCount { 0 };
```
Seems a little strange to use `size_t` for something that’s part of webpage structure, but may be the best choice. Is `uint64_t` or even `uint32_t` better?

weinig replied:
It's definitely an odd choice. uint32_t is for sure fine, since our Vectors can't get any larger than that, but changing these is a non-trivial code change.

darinadler commented on Source/WebCore/css/values/grid/CSSGridTemplateAreas.cpp:
```diff
@@ -0,0 +1,34 @@
+namespace CSS {
+
+} // namespace CSS
+} // namespace WebCore
```
Omit this source file? Or will there be code here in the future?

weinig replied:
Oops. Nope, just my error in not removing it.

darinadler commented on Source/WebCore/style/values/grid/StyleGridNamedLinesMap.cpp:
```diff
@@ -0,0 +1,39 @@
+namespace WebCore {
+namespace Style {
+
+TextStream& operator<<(TextStream& ts, const GridNamedLinesMap& value)
+{
+    return ts << value.map;
+}
```
Optional and I know we didn’t do this before, but I personally love words so much that I’d love to call these `stream` instead of `ts`.
```suggestion
TextStream& operator<<(TextStream& stream, const GridNamedLinesMap& value)
{
    return stream << value.map;
}
```

darinadler commented on Source/WebCore/style/values/grid/StyleGridTemplateAreas.cpp:
```diff
@@ -0,0 +1,115 @@
+        auto areaSpan = direction == GridTrackSizingDirection::ForRows ? area.value.rows : area.value.columns;
+        {
+            auto& startVector = namedGridLines.map.add(makeString(area.key, "-start"_s), Vector<unsigned>()).iterator->value;
+            startVector.append(areaSpan.startLine());
+            std::ranges::sort(startVector);
```
Adding a number and keeping an already-sorted vector sorted is not most efficiently done with `append` and `sort`. We should use the better algorithm instead, more like this, although this is not entirely correct:
```
template<typename ContainerType, typedef ValueType>
typename void insertSorted(ContainerType& container, ValueType&& value)
{
    ASSERT(std::ranges::is_sorted(container));
    auto location = std::ranges::upper_bound(container, value);
    return container.insert(location, std::forward<ValueType>(value));
}
```

weinig replied:
Going to leave this as is for this change. Really, I think we can just sort once at the end here, but I want to keep that kind of thing separate.

darinadler commented on Source/WebCore/style/values/grid/StyleGridTemplateAreas.h:
```diff
@@ -0,0 +1,82 @@
+        , implicitNamedGridColumnLines { }
+        , implicitNamedGridRowLines { }
+    {
+    }
```
```suggestion
    GridNamedAreaMap map { };
    GridNamedLinesMap implicitNamedGridColumnLines { };
    GridNamedLinesMap implicitNamedGridRowLines { };

    GridTemplateAreas(CSS::Keyword::None)
    {
    }
```

darinadler commented on Source/WebCore/style/values/grid/StyleGridTemplateAreas.h:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    explicit GridTemplateAreas(GridNamedAreaMap&&);
+    explicit GridTemplateAreas(const GridNamedAreaMap&);
```
Other structures like this have implicit constructors rather than `explicit`. Why a different choice here?

weinig replied:
I honestly don't remember. It made sense at the time, but now it eludes me.

weinig committed:
[Style] Convert grid-template-areas to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=295714

Reviewed by Darin Adler.

Bundles up the values for the `grid-template-areas` property into a new
`Style::GridTemplateAreas` type.

Also modernizes the CSS side a, bundling up the map and row/column count
values into a struct.

* Source/WebCore/Headers.cmake:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
* Source/WebCore/css/CSSGridTemplateAreasValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
* Source/WebCore/css/parser/CSSPropertyParserCustom.h:
* Source/WebCore/css/values/grid/CSSGridNamedAreaMap.cpp: Added.
* Source/WebCore/css/values/grid/CSSGridNamedAreaMap.h: Added.
* Source/WebCore/css/values/grid/CSSGridTemplateAreas.cpp: Added.
* Source/WebCore/css/values/grid/CSSGridTemplateAreas.h: Added.
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/style/GridArea.h:
* Source/WebCore/rendering/style/GridPositionsResolver.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StyleOrderedNamedLinesCollector.h:
* Source/WebCore/style/values/grid/StyleGridNamedLinesMap.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridNamedLinesMap.h: Added.
* Source/WebCore/style/values/grid/StyleGridOrderedNamedLinesMap.h: Added.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.h: Added.

Canonical link: https://commits.webkit.org/297304@main
